### PR TITLE
WT-13022 Add stats to track page lock wait time

### PR DESCRIPTION
**Pull request by y123456yz**
Background:
A user uses MongoDB to store core transaction data. After the business has just been migrated to MongoDB, the cluster occasionally experiences jitter issues with its slave nodes. By analyzing the diagnostic data from diagnose.data, it is determined that the problem is caused by the "rec_maximum_seconds" statistics.

### CODE REVIEW

**Comment by y123456yz** [comment]
From the above analysis, it can be seen that the disk responds quickly and there are no exceptions when performing the __wt_reconcile operation. Therefore, I suspect that the problem is that some pages have held the lock for too long during split, which ultimately leads to a longer waiting time for the page lock to be obtained by the __wt_reconcile.

**Comment by y123456yz** [comment]
https://jira.mongodb.org/browse/WT-13022

File: src/conn/conn_handle.c [hunk 1]
```diff
@@ -15,6 +15,7 @@
 int
 __wt_connection_init(WT_CONNECTION_IMPL *conn)
 {
+    WT_DECL_RET;
```
Comment (quchenhao) [inline src/conn/conn_handle.c:18]: Why this change is required?

File: src/include/mutex_inline.h [hunk 2]
```diff
@@ -341,7 +341,7 @@ __wt_spin_unlock_if_owned(WT_SESSION_IMPL *session, WT_SPINLOCK *t)
  */
 #define WT_SPIN_INIT_TRACKED(session, t, name)                                                    \
     do {                                                                                          \
-        WT_RET(__wt_spin_init(session, t, #name));                                                \
+        WT_ERR(__wt_spin_init(session, t, #name));                                                \
```
Comment (quchenhao) [inline src/include/mutex_inline.h:344]: Why you change this from RET to ERR?

File: src/include/schema.h [hunk 3]
```diff
@@ -156,6 +156,26 @@ struct __wt_import_list {
...
+ *	Acquire the btree page lock, perform an operation, drop the lock.
+ */
+#define WT_WITH_BTREE_PAGE_LOCK(session, page)                     \
+    do {                                                           \
+        __wt_spin_lock_track(session, &(page)->modify->page_lock); \
+    } while (0)
+
+#define WT_WITH_BTREE_PAGE_LOCK_NOWAIT(session, ret, page)                  \
+    do {                                                                    \
+        (ret) = 0;                                                          \
```
Comment (quchenhao) [inline src/include/schema.h:170]: This is dangerous. If we see an error before, should we have returned already?

File: src/btree/bt_split.c [hunk 4]
```diff
@@ -1191,13 +1192,14 @@ __split_internal_lock(WT_SESSION_IMPL *session, WT_REF *ref, bool trylock, WT_PA
...
 
-        if (trylock)
-            WT_RET(WT_PAGE_TRYLOCK(session, parent));
-        else
-            WT_PAGE_LOCK(session, parent);
+        if (trylock) {
+            WT_WITH_BTREE_PAGE_LOCK_NOWAIT(session, ret, parent);
+            WT_RET(ret);
+        } else
+            WT_WITH_BTREE_PAGE_LOCK(session, parent);
```
Comment (quchenhao) [inline src/btree/bt_split.c:1199]: should you handle the ret value here as well?

**Comment by quchenhao** [comment]
@y123456yz Looks like this is an debug change. I don't think we should merge this.

**Comment by quchenhao** [comment]
For time spent in reconciliation, we now track them in the reconciliation_timeline structure. How does the data inside the structure look like? I think you can add more stats for the structure to get more information where we spend time in reconciliation.

**Comment by y123456yz** [comment]
@quchenhao 
I think even without this jitter issue, it would be meaningful to add a page lock time statistics here, including the impact on internal thread and the app threads. The significance of these statistics is no less than that of other lock statistics below.
image

Because the split process itself may be time-consuming.

File: src/include/schema.h [hunk 5]
```diff
@@ -156,6 +156,25 @@ struct __wt_import_list {
     WT_WITH_LOCK_NOWAIT(                                 \
       session, ret, &S2C(session)->checkpoint_lock, WT_SESSION_LOCKED_CHECKPOINT, op)
 
+/*
```
Comment (quchenhao) [inline src/include/schema.h:159]: What about the old WT_PAGE_LOCK macro?

File: src/conn/conn_handle.c [hunk 6]
```diff
@@ -90,6 +91,7 @@ __wt_connection_init(WT_CONNECTION_IMPL *conn)
     conn->ckpt_time_min = UINT64_MAX;
     conn->ckpt_scrub_min = UINT64_MAX;
 
+err:
```
Comment (quchenhao) [inline src/conn/conn_handle.c:94]: should the return value be return (ret)?

**Comment by y123456yz** [comment]
> @y123456yz For time spent in reconciliation, we now track them in the reconciliation_timeline structure. How does the data inside the structure look like? I think you can add more stats for the structure to get more information where we spend time in reconciliation.



> @y123456yz For time spent in reconciliation, we now track them in the reconciliation_timeline structure. How does the data inside the structure look like? I think you can add more stats for the structure to get more information where we spend time in reconciliation.

Yes, I am reproducing this issue. In fact, only by adding the configuration of __ wt_timing_stress_config, such as split 5 and split 6, can this problem be reproduced immediately.

I am currently using mongodb for offline replication. The root cause of the problem，I will submit the code in another PR.  now I am not sure yet, Because I haven't verified yet. Therefore, I have not submitted the another PR yet. After confirming and verifying, I will submit another PR.

thanks.

File: src/include/btmem.h [hunk 7]
```diff
@@ -450,9 +450,6 @@ struct __wt_page_modify {
     bool instantiated;        /* True if this is a newly instantiated page. */
     WT_UPDATE **inst_updates; /* Update list for instantiated page with unresolved truncate. */
 
-#define WT_PAGE_LOCK(s, p) __wt_spin_lock((s), &(p)->modify->page_lock)
-#define WT_PAGE_TRYLOCK(s, p) __wt_spin_trylock((s), &(p)->modify->page_lock)
-#define WT_PAGE_UNLOCK(s, p) __wt_spin_unlock((s), &(p)->modify->page_lock)
```
Comment (quchenhao) [inline src/include/btmem.h:455]: @y123456yz  I don't understand why we don't modify these macros instead we need to create a set of new?

**Comment by quchenhao** [comment]
@y123456yz Can you first answer this question?

> I don't understand why we don't modify these macros instead we need to create a set of new?

I cannot find your reply may be I'm missing something?

**Comment by y123456yz** [comment]
> @y123456yz Can you first answer this question?
> 
> > I don't understand why we don't modify these macros instead we need to create a set of new?
> 
> I cannot find your reply may be I'm missing something?

I'm sorry, I didn't get what you meant.
You mean why not use WT_PAGE_LOCK here instead of creating a new WT_WITH_BTREE_PAGE_LOCK?
The reason I'm creating a new WT_WITH_BTREE_PAGE_LOCK macro here is to keep the code style consistent with the other spin_lock statistics.

**Comment by quchenhao** [comment]
> > @y123456yz Can you first answer this question?
> > > I don't understand why we don't modify these macros instead we need to create a set of new?
> > 
> > 
> > I cannot find your reply may be I'm missing something?
> 
> @quchenhao I'm sorry, I didn't get what you meant. You mean why not use WT_PAGE_LOCK here instead of creating a new WT_WITH_BTREE_PAGE_LOCK? The reason I'm creating a new WT_WITH_BTREE_PAGE_LOCK macro here is to keep the code style consistent with the other spin_lock statistics.

No they have a different meaning. The macros with "WITH" locks the lock and then unlocks it, which is a different usage type. It means the "op" we do must hold the lock.

```
/*
 * WT_WITH_LOCK_WAIT --
 *	Wait for a lock, perform an operation, drop the lock.
 */
#define WT_WITH_LOCK_WAIT(session, lock, flag, op)    \
    do {                                              \
        if (FLD_ISSET(session->lock_flags, (flag))) { \
            op;                                       \
        } else {                                      \
            __wt_spin_lock_track(session, lock);      \
            FLD_SET(session->lock_flags, (flag));     \
            op;                                       \
            FLD_CLR(session->lock_flags, (flag));     \
            __wt_spin_unlock(session, lock);          \
        }                                             \
    } while (0)
```

In the page lock usage, we lock and unlock separately which doesn't fit this pattern.

**Comment by y123456yz** [comment]
@quchenhao 
Thanks to the guidance, I have restored the macro definition here.
thanks.

File: src/include/schema.h [hunk 8]
```diff
@@ -156,6 +156,25 @@ struct __wt_import_list {
...
+ * WT_PAGE_LOCK, WT_PAGE_TRYLOCK --
+ *	Acquire the page lock
+ */
+#define WT_PAGE_LOCK(session, page)                                \
```
Comment (quchenhao) [inline src/include/schema.h:163]: Can we move this back to btmem.h? Also can we simplify to:

```
#define WT_PAGE_LOCK(s, p) __wt_spin_lock_track((s), &(p)->modify->page_lock)
#define WT_PAGE_TRYLOCK(s, p) __wt_spin_trylock_track((s), &(p)->modify->page_lock)
#define WT_PAGE_UNLOCK(s, p) __wt_spin_unlock((s), &(p)->modify->page_lock)
```

**Review by quchenhao — APPROVED** [review]
lgtm
